chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,109 +37,245 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
+    for (auto it = it_start; it != free_bins_.end(); ++it, ++distance) {
+        FreeList& free_list = it->second;
+        if (free_list.empty()) {
+            continue;
+        }
+        std::unique_ptr<Chunk> chunk = std::move(free_list.back());
+        CHAINERX_ASSERT(chunk != nullptr);
+        free_list.pop_back();
+        if (distance > kCompactionThreashold) {
+            CompactFreeBins(it_start, it);
+        }
+        return chunk;
+    }
+
+    return nullptr;
+}
+
+// Removes a chunk from an appropriate free list, and returns the removed chunk
+//
+// Not thread-safe
+std::unique_ptr<Chunk> MemoryPool::RemoveChunkFromFreeList(Chunk* chunk) {
+    CHAINERX_ASSERT(chunk != nullptr);
+
+    // Find an appropriate free list
+    auto free_bins_it = free_bins_.find(chunk->bytesize());
+    if (free_bins_it == free_bins_.end()) {
+        return nullptr;
+    }
+    FreeList& free_list = free_bins_it->second;
+
+    // Remove the given chunk from the found free list
+    for (auto it = free_list.begin(); it != free_list.end(); ++it) {
+        if (it->get() == chunk) {
+            std::unique_ptr<Chunk> chunk = std::move(*it);
+            CHAINERX_ASSERT(chunk != nullptr);
+            free_list.erase(it);
+            return chunk;
+        }
+    }
+    return nullptr;
+}
+
 MemoryPool::~MemoryPool() {
     // NOTE: CudaSetDeviceScope is not available at dtor because it may throw
     int orig_device_index{0};
     cudaGetDevice(&orig_device_index);
     cudaSetDevice(device_index_);
 
-    for (const std::vector<void*>& free_list : free_bins_) {
-        for (void* ptr : free_list) {
-            allocator_->Free(ptr);
+    for (auto& item : free_bins_) {
+        FreeList& free_list = item.second;
+        for (const std::unique_ptr<Chunk>& chunk : free_list) {
+            if (chunk->prev() == nullptr) {
+                allocator_->Free(chunk->ptr());
+            }
         }
     }
     // Ideally, in_use_ should be empty, but it could happen that shared ptrs to memories allocated
     // by this memory pool are released after this memory pool is destructed.
     // Our approach is that we anyway free CUDA memories held by this memory pool here in such case.
     // Operators of arrays holding such memories will be broken, but are not supported.
     for (const auto& item : in_use_) {
-        void* ptr = item.first;
-        allocator_->Free(ptr);
+        const std::unique_ptr<Chunk>& chunk = item.second;
+        if (chunk->prev() == nullptr) {
+            allocator_->Free(chunk->ptr());
+        }
     }
 
     cudaSetDevice(orig_device_index);
 }
 
 void MemoryPool::FreeUnusedBlocks() {
     CudaSetDeviceScope scope{device_index_};
-    std::lock_guard<std::mutex> lock{free_bins_mutex_};
-    for (const std::vector<void*>& free_list : free_bins_) {
-        for (void* ptr : free_list) {
-            allocator_->Free(ptr);
+
+    for (auto free_bins_it = free_bins_.begin(); free_bins_it != free_bins_.end();) {
```

Comment:
How about performing compaction in a separate loop? That way I think we can make use of range-based for loop, making the code much easier to read.

Reply:
It may not be good to compact `free_bins_` by erasing empty lists, because it costs O(n^2) time complexity to compact each free list (n is the length of free list).
I think it is better to make new free bins and replace older one with it.

Reply:
Hmm, sorry what do you mean? What I meant was to write this after the current loop.

```
std::remove_if(
    free_bins_.begin(),
    free_bins_.end(),
    [](const auto& pair) { return pair.second.empty(); });
```

Reply:
`std::remove_if` cannot erase the elements of `std::map` because `std::pair<const Key, T>` is not move-assignable type.

Reply:
`std::remove_if` cannot be used, but range-based-for loop can be used.  I fixed.